main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -1804,6 +1815,72 @@ object Weeder {
 
   }
 
+  /**
+    * Returns a restrictable choice rule from an already visited pattern, guard, and body.
+    * It is checked that
+    *
+    * - The guard is absent
+    *
+    * - The patterns are only tags with possible terms of variables and wildcards
+    *
+    * - The body is a direct tag if the choose is of star variant
+    */
+  private def createRestrictableChoiceRule(star: Boolean, p0: WeededAst.Pattern, g0: Option[WeededAst.Expression], b0: WeededAst.Expression): Validation[WeededAst.RestrictableChoiceRule, WeederError] = {
+    // Check that guard is not present
+    val gVal = g0 match {
+      case Some(ginner) =>
+        WeederError.UnsupportedRestrictableChooseRule(
+          s"Choose${if (star) "*" else ""} cases do not allow guards,",
+          "",
+          ginner.loc).toFailure
+      case None => ().toSuccess
+    }
+    // Check that bodies are tags for choose*
+    val bVal = b0 match {
+      case _ if !star => b0.toSuccess
+      case Expression.Apply(Expression.Ambiguous(_, _), _, _) => b0.toSuccess
```

Comment:
I think this allows `def` etc. too. Probably this check is best left to the Resolver.